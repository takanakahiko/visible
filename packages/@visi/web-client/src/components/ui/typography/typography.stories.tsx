/* cspell: disable */
import React from 'react';

import { Typography } from '.';

export default {
  title: 'Typography',
  component: Typography,
};

export const Heading1 = () => <Typography variant="h1">Text</Typography>;
export const Heading2 = () => <Typography variant="h2">Text</Typography>;
export const Heading3 = () => <Typography variant="h3">Text</Typography>;
export const Heading4 = () => <Typography variant="h4">Text</Typography>;
export const Heading5 = () => <Typography variant="h5">Text</Typography>;
export const Heading6 = () => <Typography variant="h6">Text</Typography>;

export const Paragraph = () => (
  <Typography variant="p">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
    tempor incididunt ut labore et dolore magna aliqua.
  </Typography>
);

export const Code = () => <Typography variant="code">hello world</Typography>;

export const Mixed = () => (
  <>
    <Typography variant="h4">Hello world</Typography>
    <Typography variant="p">
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
      <Typography variant="code">tempor</Typography>
      incididunt ut labore et dolore magna aliqua.
    </Typography>
  </>
);
