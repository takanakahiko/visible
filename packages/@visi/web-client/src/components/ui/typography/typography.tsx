import classNames from 'classnames';
import React from 'react';

export type TypographyVariant =
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'p'
  | 'code';

export type TypographyFontStyle = 'italic' | 'normal';
export type TypographyFontWeight = 'bold' | 'normal';
export type TypographyAlign = 'center';
export type TypographyFontSize =
  | '6xl'
  | '5xl'
  | '4xl'
  | '3xl'
  | '2xl'
  | 'xl'
  | 'lg'
  | 'base'
  | 'sm'
  | 'xs';
export type TypographyColor = 'wash' | 'normal' | 'invert';

export interface TypographyProps {
  variant: TypographyVariant;
  children: React.ReactNode;
  color?: TypographyColor;
  fontStyle?: TypographyFontStyle;
  fontWeight?: TypographyFontWeight;
  fontSize?: TypographyFontSize;
  align?: TypographyAlign;
  className?: string;
}

const mapFontSizeToClassName = (fontSize: TypographyFontSize) => {
  switch (fontSize) {
    case '6xl':
      return 'text-6xl';
    case '5xl':
      return 'text-5xl';
    case '4xl':
      return 'text-4xl';
    case '3xl':
      return 'text-3xl';
    case '2xl':
      return 'text-2xl';
    case 'xl':
      return 'text-xl';
    case 'lg':
      return 'text-lg';
    case 'sm':
      return 'text-sm';
    case 'xs':
      return 'text-xs';
    case 'base':
    default:
      return 'text-base';
  }
};

const mapAlign = (align: TypographyAlign) => {
  switch (align) {
    case 'center':
      return 'text-center';
  }
};

const inferFontSizeFromVariant = (
  variant: TypographyVariant,
): TypographyFontSize => {
  switch (variant) {
    case 'h1':
      return '6xl';
    case 'h2':
      return '5xl';
    case 'h3':
      return '4xl';
    case 'h4':
      return '3xl';
    case 'h5':
      return '2xl';
    case 'h6':
      return 'xl';
    case 'code':
      return 'sm';
    case 'p':
    default:
      return 'base';
  }
};

const mapColorPropToClassName = (color: TypographyColor) => {
  switch (color) {
    case 'wash':
      return 'text-gray-700';
    case 'invert':
      return 'text-white';
    case 'normal':
    default:
      return 'text-black';
  }
};

const isHeading = (variant: TypographyVariant) => /^h/.test(variant);

export const Typography = (props: TypographyProps) => {
  const {
    color,
    variant,
    fontSize,
    fontStyle,
    fontWeight,
    children,
    align,
    className: _class,
  } = props;

  const className = classNames(
    {
      italic: fontStyle === 'italic',
      'font-bold': isHeading(variant) || fontWeight === 'bold',
      'font-mono': variant === 'code',
    },
    variant === 'code' && [
      'rounded',
      'bg-gray-300',
      'p-1',
      // 'mx-2',
      'text-indigo-900',
    ],
    mapFontSizeToClassName(fontSize ?? inferFontSizeFromVariant(variant)),
    color && mapColorPropToClassName(color),
    align && mapAlign(align),
    _class,
  );

  return React.createElement(variant, { className }, children);
};

Typography.defaultProps = {
  variant: 'p',
};
