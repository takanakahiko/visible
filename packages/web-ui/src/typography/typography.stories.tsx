import { text } from '@storybook/addon-knobs';
import React from 'react';

import { Typography } from '.';

export default {
  title: 'Typography',
  component: Typography,
};

export const h1 = () => (
  <Typography variant="h1">{text('Text', 'Hello')}</Typography>
);

export const h2 = () => (
  <Typography variant="h2">{text('Text', 'Hello')}</Typography>
);

export const h3 = () => (
  <Typography variant="h3">{text('Text', 'Hello')}</Typography>
);

export const h4 = () => (
  <Typography variant="h4">{text('Text', 'Hello')}</Typography>
);

export const h5 = () => (
  <Typography variant="h5">{text('Text', 'Hello')}</Typography>
);

export const h6 = () => (
  <Typography variant="h6">{text('Text', 'Hello')}</Typography>
);

export const body = () => (
  <Typography variant="body">{text('Text', 'Hello')}</Typography>
);

export const italic = () => (
  <Typography fontStyle="italic">{text('Text', 'Hello')}</Typography>
);

export const wash = () => (
  <Typography color="wash">{text('Text', 'Hello')}</Typography>
);

export const uppercase = () => (
  <Typography textTransform="uppercase">{text('Text', 'Hello')}</Typography>
);
