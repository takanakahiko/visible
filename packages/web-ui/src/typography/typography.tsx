import React from 'react';
import styled from 'styled-components';

export const H1 = styled.h1`
  font-size: 28px;
  font-weight: bold;
`;

export const H2 = styled.h2`
  font-size: 24px;
  font-weight: bold;
`;

export const H3 = styled.h3`
  font-size: 21px;
  font-weight: bold;
`;

export const H4 = styled.h4`
  font-size: 18px;
  font-weight: bold;
`;

export const H5 = styled.h5`
  font-size: 16px;
  font-weight: bold;
`;

export const H6 = styled.h6`
  font-size: 14px;
  font-weight: bold;
`;

export const Body = styled.span`
  font-size: 14px;
`;

export interface TypographyProps {
  variant: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'body';
  children: React.ReactNode;
}

export const Typography = (props: TypographyProps) => {
  const { variant, children, ...rest } = props;

  switch (variant) {
    case 'h1':
      return <H1 {...rest}>{children}</H1>;
    case 'h2':
      return <H2 {...rest}>{children}</H2>;
    case 'h3':
      return <H3 {...rest}>{children}</H3>;
    case 'h4':
      return <H4 {...rest}>{children}</H4>;
    case 'h5':
      return <H5 {...rest}>{children}</H5>;
    case 'h6':
      return <H6 {...rest}>{children}</H6>;
    case 'body':
      return <Body {...rest}>{children}</Body>;
    default:
      return null;
  }
};
