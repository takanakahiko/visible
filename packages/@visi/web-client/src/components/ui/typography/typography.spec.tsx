import { render } from '@testing-library/react';
import React from 'react';

import { Mixed } from './typography.stories';

describe('Typography', () => {
  it('matches snapshot', () => {
    const { container } = render(<Mixed />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
