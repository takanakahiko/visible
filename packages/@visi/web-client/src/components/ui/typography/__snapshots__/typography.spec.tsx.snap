// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography matches snapshot 1`] = `
<h4
  class="font-bold text-3xl"
>
  Hello world
</h4>
`;
